modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/util/StreamInterceptorsLoader.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.synapse.transport.passthru.util;
+
+import org.apache.axiom.om.OMElement;
+import org.apache.axiom.om.OMNode;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.commons.util.MiscellaneousUtil;
+import org.apache.synapse.commons.util.PropertyHelper;
+import org.apache.synapse.transport.passthru.StreamInterceptor;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import javax.xml.namespace.QName;
+
+public class StreamInterceptorsLoader {
+
+    StreamInterceptorsLoader() {
+    }
+
+    private static Log log = LogFactory.getLog(StreamInterceptorsLoader.class);
+
+    private static final QName ROOT_Q = new QName("interceptors");
+    private static final QName INTERCEPTOR_Q = new QName("interceptor");
+    private static final QName CLASS_Q = new QName("class");
+    private static final QName NAME_ATT = new QName("name");
+    private static final QName PARAM_Q = new QName("parameter");
+    private static final QName VALUE_ATT = new QName("value");
+
+    private static List<StreamInterceptor> interceptors = new ArrayList<>();
+
+    private static boolean isLoadedAlready = false;
+
+    /**
+     * Load and get all synapse interceptors
+     *
+     * @return List of loaded synapse interceptors
+     */
+    public static List<StreamInterceptor> getInterceptors() {
+
+        if (!isLoadedAlready) {
```

Comment:
PassthruHttpListener.init being called concurrently? If so, this method get called concurrently and the list of interceptors will be inconsistent.

Reply:
used a lock around this